test-clients/src/main/java/com/hedera/services/bdd/suites/contract/traceability/SidecarWatcher.java

```diff
@@ -0,0 +1,195 @@
+                final var ev = (WatchEvent<Path>) event;
+                final var filename = ev.context();
+                final var child = path.resolve(filename);
+                log.info("Record stream file created -> {} ", child.getFileName());
+                final var newFilePath = child.toAbsolutePath().toString();
+                if (SIDECAR_FILE_REGEX.matcher(newFilePath).find()) {
+                    log.info("We have a new sidecar.");
+                    final var pair = RecordStreamingUtils.readSidecarFile(newFilePath);
+                    if (pair.isEmpty()) {
+                        log.fatal(
+                                "An invalid sidecar file was generated from the consensus node."
+                                        + " This is very bad.");
+                        return;
+                    }
+                    onNewSidecarFile(pair.get());
+                    if (shouldTerminateAfterNext) {
+                        watchService.close();
+                        return;
+                    }
+                }
+            }
+            // Reset the key -- this step is critical if you want to
+            // receive further watch events.  If the key is no longer valid,
+            // the directory is inaccessible so exit the loop.
+            final var valid = key.reset();
+            if (!valid) {
+                log.fatal("Error occurred in WatchServiceAPI. Exiting now.");
+                break;
+            }
+        }
+    }
+
+    private void onNewSidecarFile(final SidecarFile sidecarFile) {
+        for (final var actualSidecar : sidecarFile.getSidecarRecordsList()) {
+            if (hasSeenFirst) {
+                assertIncomingSidecar(actualSidecar);
+            } else {
+                // sidecar records from different suites can be present in the sidecar
+                // files before our first expected sidecar so skip sidecars until we reach
+                // the first expected one in the queue
+                if (expectedSidecars.isEmpty()) {
+                    continue;
+                }
+                if (expectedSidecars
+                        .peek()
+                        .getValue()
+                        .getConsensusTimestamp()
+                        .equals(actualSidecar.getConsensusTimestamp())) {
+                    hasSeenFirst = true;
+                    assertIncomingSidecar(actualSidecar);
+                }
+            }
+        }
+    }
+
+    private void assertIncomingSidecar(final TransactionSidecarRecord actualSidecar) {
+        // there should always be an expected sidecar at this point;
+        // if a NPE is thrown here, the specs have missed a sidecar
+        // and must be updated to account for it
+        final var expectedSidecarPair = expectedSidecars.poll();
+        final var expectedSidecar = expectedSidecarPair.getValue();
+
+        if ((actualSidecar.hasBytecode() || actualSidecar.hasStateChanges())
```

Comment:
Open a ticket to capture enhancements around additional record file validations:
* Verify signature file hash signed by node's public key
* Verify signature metadata hash signed by node's public key
* Verify running hash chain unbroken
* Verify record file hash matches signature hash
* Verify record file metadata hash matches signature metadata hash
* Verify sidecar hash matches corresponding sidecar hash in record file
* Verify balances (hbar, token, nft) before plus transfers in record file matches balances after

Reply:
I think I have POC code (from when we were introducing RSV6, 2-3 months ago) using `WatchService` and collecting all transactions as we send them to the node from the testing framework and making sure they are externalized in the record files and in correct order (also validating the hash chain, block number, etc.). Maybe we can dig it out and update it with everything here.

@stoqnkpL

Reply:
Opened #3823